javaslang/src/main/java/javaslang/collection/BitMappedTrie.java

```diff
@@ -246,6 +246,34 @@ void setCurrentArray() {
         };
     }
 
+    @SuppressWarnings("unchecked")
+    <T2> void visit(LeafVisitor<T2[]> visitor) {
+        int start = lastDigit(offset);
+        for (int index = 0; index < length; ) {
+            final T2[] leaf = (T2[]) getLeaf(index);
+            final int end = Math.min(leaf.length, start + length - index);
+
+            visitor.visit(leaf, start, end);
+
+            index += end - start;
+            start = 0;
+        }
+    }
+
+    <U> BitMappedTrie<U> map(Function<? super T, ? extends U> mapper) {
```

Comment:
could be optimized to take advantage of the fast that map keeps the original structure (i.e. almost `2x` speedup), but would result in slightly more complicated code.
@danieldietrich?